File: frontend/src/app/core/services/comitente.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Comitente, ComitenteRequest } from '../models';
import { Observable } from 'rxjs';
import { API } from '../constants/api.constants';

@Injectable({
  providedIn: 'root',
})
export class ComitenteService {
  private readonly http = inject(HttpClient);

  /** Lista todos los comitentes del profesional autenticado. */
  listar(): Observable<Comitente[]> {
    return this.http.get<Comitente[]>(API.COMITENTES);
  }

  /** Obtiene un comitente por ID. */
  obtenerPorId(id: number): Observable<Comitente> {
    return this.http.get<Comitente>(API.COMITENTE_BY_ID(id));
  }

  /**
   * Búsqueda incremental por DNI/CUIT. El backend ignora guiones y demás
   * separadores, y devuelve lista vacía con menos de 3 dígitos.
   */
  buscarIncremental(fragmento: string): Observable<Comitente[]> {
    const params = new HttpParams().set('q', fragmento);
    return this.http.get<Comitente[]>(API.COMITENTE_BUSCAR_INCREMENTAL, { params });
  }

  crear(request: ComitenteRequest): Observable<Comitente> {
    return this.http.post<Comitente>(API.COMITENTES, request);
  }

  actualizar(id: number, request: ComitenteRequest): Observable<Comitente> {
    return this.http.put<Comitente>(API.COMITENTE_BY_ID(id), request);
  }

  /** Soft delete (el backend marca deleted_at). */
  eliminar(id: number): Observable<void> {
    return this.http.delete<void>(API.COMITENTE_BY_ID(id));
  }
  
}

File: backend/src/main/java/ar/edu/utn/frc/previsar/services/ComitenteService.java
package ar.edu.utn.frc.previsar.services;

import ar.edu.utn.frc.previsar.dtos.request.ComitenteRequestDto;
import ar.edu.utn.frc.previsar.dtos.response.ComitenteResponseDto;

import java.util.List;
import java.util.Optional;

/**
 * Operaciones sobre la cartera de Comitentes del profesional autenticado.
 *
 * Todas las operaciones se aplican implícitamente al profesional que hace
 * la request; nunca se pasa profesionalId en los métodos. Esto evita que
 * un profesional pueda operar sobre la cartera de otro.
 */

public interface ComitenteService {
    /**
     * Lista todos los comitentes activos (no eliminados) del profesional actual.
     */
    List<ComitenteResponseDto> listar();

    /**
     * Devuelve un comitente específico por id. Falla si no pertenece al
     * profesional actual o si está eliminado.
     */
    ComitenteResponseDto obtenerPorId(Long id);

    /**
     * Busca un comitente activo por DNI/CUIT en la cartera del profesional
     * actual. Devuelve vacío si no existe (lo cual indica al frontend que
     * puede crear uno nuevo).
     */
    Optional<ComitenteResponseDto> buscarPorDniCuit(String dniCuit);

    /**
     * Búsqueda incremental por DNI/CUIT dentro de la cartera propia.
     *
     * Normaliza el fragmento (ignora guiones, puntos y espacios) y lo compara
     * contra el DNI/CUIT también normalizado, porque los DNI se guardan sin
     * separadores y los CUIT con ellos. Devuelve lista vacía con menos de 3
     * dígitos, para no traer media cartera en cada tecla.
     */
    List<ComitenteResponseDto> buscarPorFragmentoDniCuit(String fragmento);

    /**
     * Crea un nuevo comitente en la cartera del profesional actual.
     */
    ComitenteResponseDto crear(ComitenteRequestDto request);

    /**
     * Modifica un comitente existente del profesional actual.
     */
    ComitenteResponseDto actualizar(Long id, ComitenteRequestDto request);

    /**
     * Soft delete: marca el comitente como eliminado sin borrarlo físicamente.
     * Preserva la integridad de expedientes históricos.
     */
    void eliminar(Long id);
}
